### DISCUSSION TITLE
Runtime env vars?

### DISCUSSION THREAD

**Discussion by dmudro** [vercel/community#1032] [Help] [answered]
my Vercel hosted Node.js backend connects to a 3rd party API where the authorisation follows server to server OAuth flow. the issue is that access token required for the backend to backend API calls is _short lived_ so I can't just store it in an environmental variable which I conceptualise as _build time_ vars.

I reckon the underlying question here is if there is any mechanism in Vercel for my app to read some kind of _runtime_ env vars. I thought about using Vercel API to inject refreshed access token into build artifacts but these are immutable.

the current solution is (obviously) storing tokens externally which requires a round trip to database in each serverless function making the 3rd party API call (1. fetch access token from dB 2. attach the access token to the API call).

would I be able to use [secrets API](https://vercel.com/docs/rest-api#endpoints/secrets/get-a-single-secret)? but then each serverless function would have to make a REST roundtrip to Vercel instead of DB.

any ideas please for the runtime solution?

**Comment by amyegan** [comment]
Thanks for reaching out @dmudro! All environment variables are made available during deployment, so there isn't really a way to change them later. You could update an environment variable and [redeploy](https://vercel.com/docs/concepts/deployments/managing-deployments#redeploy-a-project) to include the new value.

The other option involves API calls like you're already doing. If you wanted to store the value with Vercel instead of your DB, you could use Vercel API to [edit](https://vercel.com/docs/rest-api#endpoints/projects/edit-an-environment-variable) and [retrieve environment variables](https://vercel.com/docs/rest-api#endpoints/projects/retrieve-the-environment-variables-of-a-project-by-id-or-name), which still requires a REST roundtrip as you mentioned.

**Reply by dmudro** [reply]
thanks @amyegan 

env vars available @ deployment is gold standard and work in majority of cases. any 'dynamic' vars such as auth tokens with expiry need another type of management as redeployment is not an option.

perhaps Vercel can see this as a high level feature request?

for my own case I'll compare performance of Vercel REST roundtrip vs DB for the env var fetch.

**Reply by ethomson** [reply]
Can you give me an estimate on what you mean when you say _short lived_? Are you recycling auth tokens every hour? Day? Month?

**Reply by dmudro** [reply]
the exact use case is auth'd API connection between Vercel and a 3rd party in our Next.js app. this particular API uses OAuth 2 and the access token expires after 1 hour.

hence the current pattern mentioned earlier: 1. store long lived refresh token in env vars 2. store short lived access token in a DB. 3. round trip to DB for access token

while I consider OAuth for cloud to cloud API integration as suboptimal, the reality is that these services exist and won't go away.

as a side note, there might be a way to tackle this use case on system design level and use something like a proxy or API gateway to negotiate the tokens outside of the Vercel hosted app.

anyway I thought going beyond deploy time env vars might be an interesting feature to consider and covering more use cases potentially.

**Comment by dmudro** [comment, marked as answer]
this just came out and it's almost what I need:

https://vercel.com/blog/edge-config-ultra-low-latency-data-at-the-edge

**Comment by dmudro** [comment]
I think the solution for my my specific OAuth case has been there the whole time.

problem recap:

> the current solution is (obviously) storing tokens externally which requires a round trip to database in each serverless function making the 3rd party API call (1. fetch access token from dB 2. attach the access token to the API call).

a simple solution avoiding the db rountrips upon each serverless invocation si to piggy back on [AWS Lamda global scope](https://docs.aws.amazon.com/lambda/latest/operatorguide/global-scope.html):

> Global variables retain their value between invocations in the same execution environment.

creating a simple closure should do the trick as demonstrated by following pseudocode:

```
/** memoised token will be shared across serverless instances */
let memoisedAccessToken;

// refreshToken is long lived and can be stored as env var, no issues there
async function getAccessToken(refreshToken) {
...
  if (!memoisedAccessToken || hasExpired(memoisedAccessToken)) {
    // slight concern about possible race condition, probably manageable though
    memoisedAccessToken = await negotiateNewAccessToken(refreshToken);
  }

  return memoisedAccessToken;
}
```

the limit is cold start which will always go to DB but this is still much better than going to DB for access token with each serverless invocations.

@amyegan @ethomson any thoughts please? I feel like the AWS Lamda global scope is not not well understood / documented in Vercel: https://vercel.com/guides?topics=serverless 

(my previous comment around exploration of new / upcoming edge config remains valid and I am keen to look into that once available).
